### PR TITLE
More optimiser rules

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

File: libevmasm/ExpressionClasses.cpp [hunk 1]
```diff
@@ -245,22 +245,24 @@ Rules::Rules()
 
 		// invariants involving known constants
 		{{Instruction::ADD, {X, 0}}, [=]{ return X; }},
+		{{Instruction::SUB, {X, 0}}, [=]{ return X; }},
```
Comment (chriseth) [inline libevmasm/ExpressionClasses.cpp:248]: Please add a test for this to be sure that it means `x-0` and not `0-x`.

**Review by chriseth — CHANGES_REQUESTED** [review]

**Comment by axic** [comment]
@chriseth added, hopefully?

@pirapira I see you had your own branch of optimiser rules, are any of those ready to be merged in to here?

**Comment by pirapira** [comment]
Maybe `{{Instruction::DIV, {0, X}}, [=]{ return u256(0); }},`

Is it worth adding lengthier rules like
`{{Instruction::XOR, {{{X}, {Instruction::XOR, {X, Y}}}}}, [=]{ return Y; }},`?

**Comment by axic** [comment]
So it seems like the parameter order is not enforced?

```
/home/travis/build/ethereum/solidity/test/libsolidity/SolidityOptimizer.cpp(128): error in "cse_sub_zero": check { _expectation.begin(), _expectation.end() } == { output.begin(), output.end() } failed. 

Mismatch in a position 0:  PUSH 5 !=  PUSH FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFB

Collections size mismatch: 3 != 1
```

**Comment by axic** [comment]
> Is it worth adding lengthier rules like
> `{{Instruction::XOR, {{{X}, {Instruction::XOR, {X, Y}}}}}, [=]{ return Y; }},`?

I would think so if the optimiser actually executes it. Is it doing longest match first or all matches and comparing the results?

**Comment by chriseth** [comment]
I don't think that the extended xor rule would be triggered often. I also never profiled the matching routine, it might be quite slow.

**Comment by chriseth** [comment]
@axic if you use constants in the tests, the optimizer will just evaluate the full expression. You need a tests like `dup1 0 sub` and `0 dup1 sub`

**Comment by axic** [comment]
@chriseth I have one more rule to inclue and perhaps some from @pirapira's changeset.

**Comment by axic** [comment]
Cherry-picked 115d30e.

Need to add tests.

**Review by chriseth — APPROVED** [review]

**Comment by axic** [comment]
Verified the optimisation for the correct case (`n - 0`).

However for the `0 - n` case it seems to be broken:

Mismatch in a position 0:  DUP1 !=  PUSH 0
Mismatch in a position 1:  PUSH 0 !=  DUP2
Mismatch in a position 2:  SUB !=  SWAP1

**Comment by axic** [comment]
It actually "optimises" `DUP1, PUSH 0, SUB` to `PUSH 0, DUP2, SWAP1, SUB`, which is correct, but not very optimised.

Too bad `checkCSE` doesn't report the difference in instruction count so it wasn't apparent.

**Review by chriseth — APPROVED** [review]

**Pull request merged**